### PR TITLE
[SPARK-45132][SQL] Fix IDENTIFIER for function invocation

### PR DESCRIPTION
**Pull request by srielau**
### What changes were proposed in this pull request?

Due to a quirk in the parser, in some cases, IDENTIFIER(<funcStr>)(<arg>) is not properly recognized as a function invocation.

The change is to remove the explicit IDENTIFIER-clause rule in the function invocation grammar and instead recognize
IDENTIFIER(<arg>) within visitFunctionCall.
  
### Why are the changes needed?

Function invocation support for IDENTIFIER is incomplete otherwise

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Added new testcases to identifier-clause.sql

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

File: sql/core/src/test/resources/sql-tests/analyzer-results/identifier-clause.sql.out [hunk 1]
```diff
@@ -187,10 +187,11 @@ Project [coalesce(cast(null as int), 1) AS coalesce(NULL, 1)#x]
 
 
 -- !query
-SELECT IDENTIFIER('abs')(-1)
+SELECT IDENTIFIER('abs')(c1) FROM VALUES(-1) AS T(c1)
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/analyzer-results/identifier-clause.sql.out:190]: what was the behavior before this PR?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 2]
```diff
@@ -2274,19 +2267,32 @@ class AstBuilder extends DataTypeAstBuilder with SQLConfHelper with Logging {
...
 
-    // Check if the function is evaluated in a windowed context.
-    ctx.windowSpec match {
-      case spec: WindowRefContext =>
-        UnresolvedWindowExpression(func, visitWindowRef(spec))
-      case spec: WindowDefContext =>
-        WindowExpression(func, visitWindowDef(spec))
-      case _ => func
+    // Is this an IDENTIFIER clause instead of a function call?
+    if (name.toLowerCase(Locale.ROOT) == "identifier" && // IDENTIFIER (but not `IDENTIFIER`)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:2272]: in order to check this more explicitly, shall we update the parser rule a bit? `functionName LEFT_PAREN ...` -> `(IDENTIFIER_KW | functionName) LEFT_PAREN`

Comment (srielau) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:2272]: @cloud-fan This doesn't work. (see commit history for what I tried).
On the bubble I think the current solution is the best.

File: sql/api/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBaseParser.g4 [hunk 3]
```diff
@@ -1196,6 +1195,7 @@ qualifiedNameList
 
 functionName
     : IDENTIFIER_KW LEFT_PAREN expression RIGHT_PAREN
+    | identFunc=IDENTIFIER_KW  // IDENTIFIER itself is also a valid function name.
```
Comment (cloud-fan) [inline sql/api/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBaseParser.g4:1198]:
```suggestion
    | identFunc=IDENTIFIER_KW   // IDENTIFIER itself is also a valid function name.
```

File: sql/api/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBaseParser.g4 [hunk 4]
```diff
@@ -1196,6 +1195,7 @@ qualifiedNameList
 
 functionName
     : IDENTIFIER_KW LEFT_PAREN expression RIGHT_PAREN
+    | identFunc=IDENTIFIER_KW   // IDENTIFIER itself is also a valid function name.
```
Comment (cloud-fan) [inline sql/api/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBaseParser.g4:1198]:
```suggestion
    | identFunc=IDENTIFIER_KW  // IDENTIFIER itself is also a valid function name.
```

File: sql/api/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBaseParser.g4 (hunk 3 above)
Comment (cloud-fan) [inline sql/api/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBaseParser.g4:1198]:
```suggestion
    | identFunc=IDENTIFIER_KW   // IDENTIFIER itself is also a valid function name.
```

**Pull request closed without merging**